cmd/cloud-controller-manager/app/controllermanager.go

```diff
@@ -113,6 +115,15 @@ the cloud specific control loops shipped with Kubernetes.`,
 
 // Run runs the ExternalCMServer.  This should never exit.
 func Run(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.TODO())
+	go func() {
+		select {
+		case <-stopCh:
+			cancel()
+		case <-ctx.Done():
```

Comment:
What is the purpose of having both a stop channel and a done channel here? Especially as a context is usually associated with a request and our main run method does not seem related to a request.

Reply:
`stop` channel is controlled by the signals( SIGTERM and SIGINT) whereas `done` channel is controlled by `Run` method and allows for graceful termination. For example, `done` channel will be closed when the component cannot create `HTTP{S}` sockets, when it loses leadership or when it receives one of the signals. Note that closing one of the channels is equivalent to requesting closing the application.